topics/RDMA.md

```diff
@@ -0,0 +1,170 @@
+
+    192.168.122.100:6379> CONFIG SET rdma-port 6380
+
+Valkey can run both RDMA and TCP concurrently on the same port:
+
+    ./src/valkey-server --protected-mode no \
+         --loadmodule src/valkey-rdma.so bind=192.168.122.100 port=6379 \
+         --port 6379
+
+Note that the network interface (192.168.122.100 of this example) should support
+RDMA. To test a server supports RDMA or not:
+
+    ~# rdma dev show (a new version iproute2 package)
+Or:
+
+    ~# ibv_devices (ibverbs-utils package of Debian/Ubuntu)
+
+
+## Protocol
+
+The protocol defines the QP type RC (like TCP), communication commands, and payload exchange mechanism.
+This dependency is based solely on the RDMA (aka Infiniband) specification
+and is independent of both software (including the OS and user libraries)
+and hardware (including vendors and low-level transports).
+
+Valkey Over RDMA has control-plane (control messages) and data-plane (payload transfer).
+
+### Control message
+
+Control messages use fixed 32-byte big-endian message structures:
+```C
+typedef struct ValkeyRdmaFeature {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    /* select features */
+    uint16_t select;
+    uint8_t rsvd[20];
+    /* feature bits */
+    uint64_t features;
+} ValkeyRdmaFeature;
+
+typedef struct ValkeyRdmaKeepalive {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    uint8_t rsvd[30];
+} ValkeyRdmaKeepalive;
+
+typedef struct ValkeyRdmaMemory {
+    /* defined as following Opcodes */
+    uint16_t opcode;
+    uint8_t rsvd[14];
+    /* address of a transfer buffer which is used to receive remote streaming data,
+     * aka 'RX buffer address'. The remote side should use this as 'TX buffer address' */
+    uint64_t addr;
+    /* length of the 'RX buffer' */
+    uint32_t length;
+    /* the RDMA remote key of 'RX buffer' */
+    uint32_t key;
+} ValkeyRdmaMemory;
+
+typedef union ValkeyRdmaCmd {
+    ValkeyRdmaFeature feature;
+    ValkeyRdmaKeepalive keepalive;
+    ValkeyRdmaMemory memory;
+} ValkeyRdmaCmd;
+```
+
+### Opcodes
+|Command| Value | Description |
+| :----: | :----: | :----: |
+| `GetServerFeature`   | 0 | required, get the features offered by Valkey server |
+| `SetClientFeature`   | 1 | required, negotiate features and set it to Valkey server |
+| `Keepalive`          | 2 | required, detect unexpected orphan connection |
+| `RegisterXferMemory` | 3 | required, tell the 'RX transfer buffer' information to the remote side, and the remote side uses this as 'TX transfer buffer' |
+
+### RDMA Operations
+- Send a control message by RDMA '**`ibv_post_send`**' with opcode '**`IBV_WR_SEND`**' with structure
+  'ValkeyRdmaCmd'.
+- Receive a control message by RDMA '**`ibv_post_recv`**', and the received buffer
+  size should be size of 'ValkeyRdmaCmd'.
+- Transfer stream data by RDMA '**`ibv_post_send`**' with opcode '**`IBV_WR_RDMA_WRITE`**' (optional) and
```

Comment:
What do you mean by '**`IBV_WR_RDMA_WRITE`**'  being optional? Is there a specific time that it's optional?

Reply:
For example, a server wants to write iovec[3] of 128, 256, 128 bytes:
* 1, 'IBV_WR_RDMA_WRITE' 128 bytes, 'IBV_WR_RDMA_WRITE' 256 bytes, 'IBV_WR_RDMA_WRITE_WITH_IMM' 128 bytes and IMMDT is 512, this means: write 512 bytes to the remote side. the remote side is awakened only one time.
* 2, IBV_WR_RDMA_WRITE_WITH_IMM' 128 bytes and IMMDT is 128, 'IBV_WR_RDMA_WRITE_WITH_IMM' 256 bytes and IMMDT is 256, 'IBV_WR_RDMA_WRITE_WITH_IMM' 128 bytes and IMMDT is 128. this means: write 512 bytes to the remote side. the remote side is awakened 3 times.

The two version write the same 512 bytes to the remote side, but 1 has better performance(more complicated to implement).